### ISSUE TITLE
Using custom Bootstrap JS modules

### ISSUE THREAD

**Issue by axlright** [understrap/understrap#1116]
Hello friends-

Are any of you compiling custom / selected Bootstrap JS modules instead of the full `bootstrap.js`? If so, mind sharing your setup? I have some clients who need super compiled / reduced everything for performance reasons.

Looking at `node_modules/bootstrap/js` there's individual modules in `dist` and in `src` directories but I'm not sure the difference between them.

I'd assume a workflow would be to modify `gulpfile.js` to pull all those `node_modules/bootstrap/js` and then modify the `scripts` function to compile certain components in place of `paths.dev + '/js/bootstrap4/bootstrap.js',`

Any input is appreciated!

-Alex

**Comment by 0dp** [comment]
hey Buddy, hope you are doing well!

Imma have a look and get right back to you

**Comment by 0dp** [comment]
There are two options right of the bat.

## Quick and dirty
Just like you write, you can import each file from `node_modules/bootstrap/js/dist` as you need them...You always need `util.js`

How you chose to do that depends on how complicated you want it, the out come will be the same.

Start by changing the copy-assets task to copy `node_modules/bootstrap/js/dist` files instead of just the bootstrap bundle and copy all the files into js folder.
```
	var stream = gulp
		.src(`${paths.node}bootstrap/js/dist/**/*.js`)
		.pipe(gulp.dest(`./js/bootstrap4`));
```

create a bootstrap.js file in ./src called boostrap for instance.
```
import './bootstrap4/util.js'
import './bootstrap4/alert.js'
import './bootstrap4/button.js';
...
```

Then modify the scripts task

```
`${paths.dev}/js/bootstrap.js`,
```

Now when you run scripts task these files will be imported, unminified, as is, no thrills.
Pretty much the same as importing them directly from the boostrap npm package. Janky af 👎 

## The right way
The right way to do this would be to setup a propper build task for this. Which you know is something I've been pushing for, for years now.

Basically create a build system that pulls in all the dependencies from our modules in compile time instead of the copy-assets task. The current approach made sense a long time ago, but not anymore.

This is what you see most modern apps do these days. I was about to go on a rant about this, but I'll leave here. IMHO to do this properly it would get political real quick

**Comment by Thomas-A-Reinert** [comment]
I am totally with you. It loads (possible) bloat.
MY personally preferred way would be an installer script that asks within initial "`npm install`" which js-modules you want. According to that it would write out some config-file, letting the gulp-task know which modules to load/compile.
Same goes for bootstrap.css by the way. I almost never use tables and such, so it would be really great to have some installer script/dialog about which css-modules you want to integrate and get rid of the rest.

**Comment by axlright** [comment]
Thanks @0dp and @Thomas-A-Reinert !!

As always, y'all are more talented than I am. I'm gonna craft up and test a more quick and dirty approach and I'll post the workflow here and let y'all know the results.

**Comment by 0dp** [comment]
It should work just fine buddy 💪

**Comment by axlright** [comment]
I tested with this in `gulpfile.js` as an edit to the `gulp scripts` action:

```
gulp.task( 'scripts', function() {
	var scripts = [

		// Start Bootstrap
		// Optionally use all Bootstrap JS components...
		// `${paths.dev}/js/bootstrap4/bootstrap.bundle.js`,

		//... or comment out above and select only components that the project uses while always including util.js

		`${paths.dev}/js/bootstrap4/util.js`, // always include
		`${paths.dev}/js/bootstrap4/alert.js`, //NAME
		`${paths.dev}/js/bootstrap4/button.js`, //NAME$1
		`${paths.dev}/js/bootstrap4/carousel.js`, //NAME$2
		`${paths.dev}/js/bootstrap4/collapse.js`, //NAME$3
		`${paths.dev}/js/bootstrap4/dropdown.js`, //NAME$4
		`${paths.dev}/js/bootstrap4/modal.js`, // NAME$5
		`${paths.dev}/js/bootstrap4/tooltip.js`, //NAME$6
		`${paths.dev}/js/bootstrap4/popover.js`, //NAME$7
		`${paths.dev}/js/bootstrap4/scrollspy.js`, //NAME$8
		`${paths.dev}/js/bootstrap4/tab.js`, //NAME$9
		`${paths.dev}/js/bootstrap4/toast.js`, // NAME$a
		`${paths.dev}/js/bootstrap4/index.js`,
		// End Bootstrap

		`${paths.dev}/js/skip-link-focus-fix.js`,

		// START CUSTOMIZED ADDITIONS - PRESERVE THROUGH UPDATES
		// Owl Carousel 2 scripts
		// paths.dev + '/js/owl.carousel.js',

		// Custom scripts
		paths.custom_js + '*.js'

		// END CUSTOMIZED ADDITIONS - PRESERVE THROUGH UPDATES
	];
  gulp.src( scripts, { allowEmpty: true } )
	.pipe( concat( 'app.min.js' ) )
	.pipe( uglify() )
	.pipe( gulp.dest( paths.js ) );

  return gulp.src( scripts, { allowEmpty: true } )
	.pipe( concat( 'app.js' ) )
	.pipe( gulp.dest( paths.js ) );
});
```

The js file it creates is smaller than that of the one using `bootstrap.bundle.js` and it throws a console error for two instances of "Uncaught TypeError: Bootstrap's JavaScript requires jQuery. jQuery must be included before Bootstrap's JavaScript." which come from this in `index.js` though everything (in my tests so far) work including collapse, dropdown and a few others.

```
if (typeof $ === 'undefined') {
    throw new TypeError('Bootstrap\'s JavaScript requires jQuery. jQuery must be included before Bootstrap\'s JavaScript.');
  }
```

Commenting out `${paths.dev}/js/bootstrap4/index.js,` makes other things unhappy.

There's not a ton of information on forums / elsewhere online about compiling a custom set of Bootstrap js components. Any thoughts? I'm gonna keep experimenting.

**Comment by 0dp** [comment]
Well the most important thing is you include the utils.js
Whatever you include after that should work. That's what I did.

I agree there is not a ton of info about this.
But try with one module at the time.
Like modal for instance. Then build on that.

You got this

**Comment by axlright** [comment]
Thanks @0dp here's what worked for me - removing the `index.js` after got rid of all console errors...

```
// Start Bootstrap
        // Optionally use all Bootstrap JS components...
        // `${paths.dev}/js/bootstrap4/bootstrap.bundle.js`,

        //... or comment out above and select only components that the project uses while always including util.js

        `${paths.dev}/js/bootstrap4/util.js`, // always include
        `${paths.dev}/js/bootstrap4/alert.js`, //NAME
        `${paths.dev}/js/bootstrap4/button.js`, //NAME$1
        `${paths.dev}/js/bootstrap4/carousel.js`, //NAME$2
        `${paths.dev}/js/bootstrap4/collapse.js`, //NAME$3
        `${paths.dev}/js/bootstrap4/dropdown.js`, //NAME$4
        `${paths.dev}/js/bootstrap4/modal.js`, // NAME$5
        `${paths.dev}/js/bootstrap4/tooltip.js`, //NAME$6
        `${paths.dev}/js/bootstrap4/popover.js`, //NAME$7
        `${paths.dev}/js/bootstrap4/scrollspy.js`, //NAME$8
        `${paths.dev}/js/bootstrap4/tab.js`, //NAME$9
        `${paths.dev}/js/bootstrap4/toast.js`, // NAME$a
        // `${paths.dev}/js/bootstrap4/index.js`,
        // End Bootstrap
```
I'll keep working project by project and maybe snoop around the Bootstrap Github to see if there's more explanation of how to compile individual modules.

**Comment by Thomas-A-Reinert** [comment]
> Thanks @0dp and @Thomas-A-Reinert !!
> 
> As always, y'all are more talented than I am.

Sorry to say, but that´s total BS ;)
I usually have a problem, no clue about how to solve it but dig into it like a mole and don´t let go like a stafford terrier. YOU solved this problem within hours, didn´t you? Don´t hide your light under a bushel, dude..

**Comment by mitchelldmiller** [comment]
Hello Nice People:
I learned programming on punch cards, never learned new NPM stuff. But this is how I got example by @axlright to work on my system. Copied Babel() from original file.
Best wishes,
Mitchell

       gulp.task( 'scripts', function() {
    	 var scripts = [
    		paths.node + '/bootstrap/js/src/util.js',
     		paths.node + '/bootstrap/js/src/alert.js',
	     	paths.node + '/bootstrap/js/src/button.js',
     		paths.node + '/bootstrap/js/src/carousel.js',
    		paths.node + '/bootstrap/js/src/collapse.js',
    		paths.node + '/bootstrap/js/src/dropdown.js',
   		paths.node + '/bootstrap/js/src/modal.js',
   		paths.node + '/bootstrap/js/src/tooltip.js',
   	   	paths.src + '/js/skip-link-focus-fix.js',
         	];
   
     gulp.src( scripts, { allowEmpty: true } )
    	.pipe( babel( { presets: ['@babel/preset-env'] } ) )
       	.pipe( concat( 'app.min.js' ) )
     	.pipe( uglify() )
      	.pipe( gulp.dest( paths.js ) );
     
       return gulp.src( scripts, { allowEmpty: true } )
     	.pipe( babel() )
     	.pipe( concat( 'app.js' ) )
     	.pipe( gulp.dest( paths.js ) );
    });
